src/auxiliary/math.jl

```diff
@@ -151,10 +151,12 @@ Given ε = 1.0e-4, we use the following algorithm.
   [https://www.agner.org/optimize/instruction_tables.pdf](https://www.agner.org/optimize/instruction_tables.pdf)
 """
 @inline function ln_mean(x, y)
+    RealT = eltype(x)
     epsilon_f2 = 1.0e-4
     f2 = (x * (x - 2 * y) + y * y) / (x * (x + 2 * y) + y * y) # f2 = f^2
     if f2 < epsilon_f2
-        return (x + y) / @evalpoly(f2, 2, 2/3, 2/5, 2/7)
+        return (x + y) / @evalpoly(f2, 2, convert(RealT, 2 / 3), convert(RealT, 2 / 5),
+                         convert(RealT, 2 / 7))
     else
```

Comment:
```julia
@inline function ln_mean_test64(x, y)
    f2 = (x * (x - 2 * y) + y * y) / (x * (x + 2 * y) + y * y)
    v1 = (x + y) / @evalpoly(f2, 2, 2/3, 2/5, 2/7)
    v2 = (y - x) / log(y / x)
    return abs(v1 - v2) / v2
end

@inline function ln_mean_test32(x, y)
    f2 = (x * (x - 2 * y) + y * y) / (x * (x + 2 * y) + y * y)
    v1 = (x + y) / @evalpoly(f2, 2, convert(Float32, 2 / 3), convert(Float32, 2 / 5),
                   convert(Float32, 2 / 7))
    v2 = (y - x) / log(y / x)
    return abs(v1 - v2) / v2
end

# Float64 (assume 0 < x < y w.l.o.g.)
# Assume x = 1.0 for simplicity, then y < (2ϵ + 2) / (1 - ϵ) and 
# try to find the relative error corresponding to ϵ = 1.0e-4
guess = 1.0e-3 # some magnitude larger than 1.0e-4
x = 1.0
y_min = 2.0
y_max = 2 + 4 * guess / (1 - guess)
step = (y_max - y_min) / 100

for y in y_min:step:y_max
    epsilon = (y - 2) / (y + 2)
    error = ln_mean_test64(x, y)
    if 0.90e-4 < epsilon < 1.10e-4 # show around 1.0e-4
        println("error = $error, epsilon = $epsilon")
    end
end

# Find the relative error is roughly in (1.795823802303315e-5, 1.7962128754169207e-5)
err_min = 1.795823802303315e-5
err_max = 1.7962128754169207e-5

# Float32 (assume 0 < x < y w.l.o.g.)
# Assume x = 1.0f0 for simplicity, then y < (2ϵ + 2) / (1 - ϵ) and try to find the best ϵ
guess = Float32(1.0e-3) # guess the upper bound of ϵ 
x = 1.0f0
y_min = 2.0f0
y_max = 2 + 4 * guess / (1 - guess)
step = (y_max - y_min) / 1000

for y in (y_min + step):step:y_max
    epsilon = (y - 2) / (y + 2)
    error_curr = ln_mean_test32(x, y)
    error_prev = ln_mean_test32(x, y - step)
    if error_prev <= err_min && error_curr >= err_max 
        # relative error for Float32 is not stable, i.e., it is not monotonically increasing with y
        println("error = $error_curr, epsilon = $epsilon")
    end
end
```
Please check above. I'm not sure how to modify the number of terms in the Taylor expansion (as it is something long time ago for me). But based on the the assumption that everything in the formula works fine for both `Float64` and `Float32`, I found that maybe `1.0e-5` is suitable for `Float32` (not pretty sure about it). I'm curious why you initially set `1.0e-4` for `Float64`. Any suggestions or useful references would be appreciated. Thanks!

Reply:
The implementation is based on https://www.math.umd.edu/~tadmor/references/files/Ismail%20and%20Roe%20affordable%20fluxes%20JCP2009.pdf
From my point of view, it's also fine to basically use the same numbers and cut-off values for `Flaot32` for now and add it to the TODO list for latr considerations when we want to optimize `Float32` computations on the GPU

Reply:
```julia
# Same definition for `ln_mean_test64` and `ln_mean_test32` (see above code)

# Float64 (assume 0 < x < y w.l.o.g.)
# Assume x = eps(Float64), then y < (2ϵ + 2)x / (1 - ϵ) and 
# try to explore the relative error when x is extremely small
guess = 1.0e-3 # some magnitude larger than 1.0e-4
x = eps(Float64)
y_min = 2 * x
y_max = (2 * guess + 2) * x / (1 - guess)
step = (y_max - y_min) / 1000

for y in y_min:step:y_max
    epsilon = (y / x - 2) / (y / x + 2)
    error = ln_mean_test64(x, y)
    if 0.90e-4 < epsilon < 1.10e-4 # show around 1.0e-4
        println("error = $error, epsilon = $epsilon")
    end
end

# Find the relative error is roughly in (1.795823802303315e-5, 1.7965630958233013e-5)
err_min = 1.795823802303315e-5
err_max = 1.7965630958233013e-5

# Float32 (assume 0 < x < y w.l.o.g.)
# Assume x = eps(Float32), then y < (2ϵ + 2)x / (1 - ϵ) and 
# try to find the corresponding ϵ that fits the relative error range
guess = Float32(1.0e-3) # guess the upper bound of ϵ 
x = eps(Float32)
y_min = 2 * x
y_max = (2 * guess + 2) * x / (1 - guess)
step = (y_max - y_min) / 1000

for y in (y_min + step):step:y_max
    epsilon = (y / x - 2) / (y / x + 2)
    error_curr = ln_mean_test32(x, y)
    error_prev = ln_mean_test32(x, y - step)
    if error_prev <= err_min && error_curr >= err_max
        # relative error for Float32 is not stable, i.e., it is not monotonically increasing with ϵ (or y)
        # but for Float64, it is monotonically increasing with ϵ (or y) please check 
        println("error = $error_curr, epsilon = $epsilon")
    end
end
```

Reply:
Thanks for the reference! I checked the Appendix B. Is there any reason you initially set epsilon to 1.0e-4 for `Float64`? As I basically do these things based on the assumption that `epsilon` value is chosen reasonably here. 

I ran the code with `x` and `y` extremely close to zero. The relative error remains unstable for `Float32`, likely due to its lower precision. If we want to control the relative error to be similar to the case of `Float64`, `epsilon` should be set to 1.0e-6 (more conservative, ensuring each result is within the relative error range). Also, 1.0e-5 is also acceptable (less conservative, covering most results within the range).

Reply:
Why did you say the original one is also acceptable now?

Reply:
> Why did you say the original one is also acceptable now?

I wanted to let you know that you don't have to do this tuning. Of course, you're free to do so if you like, but it will likely be more work since it's a mix of implementation ("make `Float32` work) and performance tuning ("make it as fast as possible in `Float32`")

Reply:
OK. I thought it was left within my PR so I was responsible for it. Do you mean it is left for you to solve or is it going to be resolved later?

Reply:
Thanks. I would be happy if you could find someone better to deal with this issue. But if no one would like to pick it up, I am okay with taking it on myself. Just ensuring that this issue potentially falls onto my project list https://github.com/users/huiyuxie/projects/2

I have already added it to the TODO list above.

Reply:
Great, thanks 👍